File: leetcode-js/41.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function (nums) {
    let currIdx = 0;
    nums.push(nums[0]);

    for (let i = 0; i < nums.length; i++) {
        nums[i] = nums[i] < 0 ? 0 : nums[i];
    }

    for (let num of nums) {
        let nextNum = num;
        while (nextNum < nums.length && nums[nextNum] >= 0) {
            let temp = nextNum;
            nextNum = nums[nextNum];
            nums[temp] = -1;
        }
    }

    let result = 1;
    while (result - 1 < nums.length && nums[result] < 0) {
        result++;
    }
    return result;
};

File: leetcode-go/41.go
package main

// import "fmt"

// my variant - solves the problem
func firstMissingPositive(nums []int) int {
	const MAX_NUM int = 100000
	checkedNums := [MAX_NUM]bool{}

	for _, num := range nums {
		if num <= MAX_NUM && num >= 1 {
			checkedNums[num-1] = true
		}
	}

	result := 1
	for result <= MAX_NUM && checkedNums[result-1] {
		result++
	}
	return result
}

// func main() {
// 	nums := []int{-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
// 	fmt.Println(firstMissingPositive(nums))
// }
